### PR TITLE
SOLR-13942: /api/cluster/zk/* to fetch raw ZK data

### PR DESCRIPTION
**Pull request by noblepaul**

### CODE REVIEW

**Comment by chatman** [comment]
This must have the reference guide changes too.

File: solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+/**Exposes the content of the Zookeeper
+ * This is an expert feature that exposes the data inside the back end zookeeper.This API may change or
+ * be removed in future versions.
+ * This is not a public API. The data that is returned is not guaranteed to remain same
+ * across releases, as the data stored in Zookeeper may change from time to time.
+ */
+@EndPoint(path = "/cluster/zk/*",
+    method = SolrRequest.METHOD.GET,
+    permission = COLL_READ_PERM)
+public class ZkRead {
```
Comment (tflobbe) [inline solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java]: The name of this class is not very representative. Also, all ZooKeeper handlers are called `ZooKeeper...`, we should keep that for consistency

File: solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+  private final CoreContainer coreContainer;
```
Comment (tflobbe) [inline solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java]: Do we need to keep the CoreContainer? is the zkClient enough?

File: solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+                ew1.put("cversion", stat.getCversion());
+                ew1.put("czxid", stat.getCzxid());
+                ew1.put("ephemeralOwner", stat.getEphemeralOwner());
+                ew1.put("mtime", stat.getMtime());
+                ew1.put("mzxid", stat.getMzxid());
+                ew1.put("pzxid", stat.getPzxid());
+                ew1.put("dataLength", stat.getDataLength());
+              });
+            } catch (Exception e) {
+              ew.put("s", Collections.singletonMap("error", e.getMessage()));
```
Comment (tflobbe) [inline solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java]: you mean `"s"` or `s`? Also, the idea here is that we'll add an error per child?

File: solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+                ew1.put("ctime", stat.getCtime());
+                ew1.put("cversion", stat.getCversion());
+                ew1.put("czxid", stat.getCzxid());
+                ew1.put("ephemeralOwner", stat.getEphemeralOwner());
+                ew1.put("mtime", stat.getMtime());
+                ew1.put("mzxid", stat.getMzxid());
+                ew1.put("pzxid", stat.getPzxid());
+                ew1.put("dataLength", stat.getDataLength());
+              });
+            } catch (Exception e) {
```
Comment (tflobbe) [inline solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java]: Why are you catching `Exception` here? instead of a subclass? Also, `exists` throws interrupted exception, you should interrupt (and break)

File: solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+        d = coreContainer.getZkController().getZkClient().getData(path, null, null, false);
+        if (d == null || d.length == 0) {
+          rsp.add(path, null);
+          return;
+        }
+
+        Map<String, String> map = new HashMap<>(1);
+        map.put(WT, "raw");
+        map.put(OMIT_HEADER, "true");
+        req.setParams(SolrParams.wrapDefaults(new MapSolrParams(map), req.getParams()));
```
Comment (tflobbe) [inline solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java]: Why is this needed?

File: solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+
+        rsp.add(CONTENT, new ContentStreamBase.ByteArrayStream(d, null,
+            d[0] == '{' ? CommonParams.JSON_MIME : BinaryResponseParser.BINARY_CONTENT_TYPE));
+
+      }
+
+    } catch (Exception e) {
```
Comment (tflobbe) [inline solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java]: Same as above, we probably don't need to catch all exceptions, just the ones being thrown. And the same thing with interruption

File: solr/core/src/test/org/apache/solr/handler/admin/ZookeeperStatusHandlerTest.java [hunk 7]
```diff
@@ -74,6 +78,39 @@ public void tearDown() throws Exception {
     super.tearDown();
   }
 
+  @Test
+  public void testZkread() throws Exception {
```
Comment (tflobbe) [inline solr/core/src/test/org/apache/solr/handler/admin/ZookeeperStatusHandlerTest.java]: This test doesn't belong in `ZookeeperStatusHandlerTest`

File: solr/core/src/test/org/apache/solr/handler/admin/ZookeeperStatusHandlerTest.java [hunk 8]
```diff
@@ -74,6 +78,39 @@ public void tearDown() throws Exception {
...
+    URL baseUrl = cluster.getJettySolrRunner(0).getBaseUrl();
+    String basezk = baseUrl.toString().replace("/solr", "/api") + "/cluster/zk";
+
+    try(  HttpSolrClient client = new HttpSolrClient.Builder(baseUrl.toString()).build()) {
```
Comment (tflobbe) [inline solr/core/src/test/org/apache/solr/handler/admin/ZookeeperStatusHandlerTest.java]: nit. Fix spaces

File: solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+                ew1.put("pzxid", stat.getPzxid());
+                ew1.put("dataLength", stat.getDataLength());
+              });
+            } catch (Exception e) {
+              ew.put("s", Collections.singletonMap("error", e.getMessage()));
+            }
+          }
+        });
+
+      } else {
```
Comment (tflobbe) [inline solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java]: Should we consider the case where znodes have data and children?

File: solr/core/src/test/org/apache/solr/handler/admin/ZookeeperStatusHandlerTest.java [hunk 10]
```diff
@@ -74,6 +78,39 @@ public void tearDown() throws Exception {
...
+          Utils.JSONCONSUMER );
+      assertNotNull(o);
+      o = Utils.executeGET(client.getHttpClient(),
+          basezk + "/configs",
+          Utils.JSONCONSUMER );
+      assertEquals("0", String.valueOf(getObjectByPath(o,true, split(":/configs:_default:dataLength",':'))));
+      assertEquals("0", String.valueOf(getObjectByPath(o,true, split(":/configs:conf:dataLength",':'))));
+      byte[] bytes = new byte[1024*5];
+      for (int i = 0; i < bytes.length; i++) {
+        bytes[i] = (byte) random().nextInt(128);
```
Comment (tflobbe) [inline solr/core/src/test/org/apache/solr/handler/admin/ZookeeperStatusHandlerTest.java]: why make things so complicated? cant you just put some static text? what are you trying to achieve with this randomization?

File: solr/core/src/test/org/apache/solr/handler/admin/ZookeeperStatusHandlerTest.java (hunk 7 above)
Comment (janhoy) [reply solr/core/src/test/org/apache/solr/handler/admin/ZookeeperStatusHandlerTest.java]: That's true. ZKStatusHandler is the handler that pulls 4LW status from ZK TCP API, which is a completely different handler :)

File: solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java (hunk 2 above)
Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java]: The zkclient may get closed or recreated. Always safer to get the current one

File: solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java (hunk 3 above)
Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java]: good catch. it was supposed to be `s`

File: solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java (hunk 4 above)
Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java]: does it even matter? 

Why interrupt the thread when this thread is not doing anything at all. This is the `qtp` thread. Interrupting it makes no sense

File: solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java (hunk 5 above)
Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java]: This is important if we are ever writing out raw content that comes from ZK. We should not use any of the `ResponseWriter` in Solr

File: solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java (hunk 6 above)
Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java]: it does not matter in this case

File: solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java (hunk 9 above)
Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java]: fixing it. you can explicitly request for the node data using a param `leaf=true`

File: solr/core/src/test/org/apache/solr/handler/admin/ZookeeperStatusHandlerTest.java (hunk 10 above)
Comment (noblepaul) [reply solr/core/src/test/org/apache/solr/handler/admin/ZookeeperStatusHandlerTest.java]: this was the easiest way to create a `byte[]` .  could use anything else. But this code was simpler

**Comment by janhoy** [comment]
Not sure I like the magic switch between returning actual data and returning a list of metdatata simply based on whether a znode happens to have children or not. This is not deterministic. I'd rather have an explicit action, e.g. `GET /api/cluster/zookeeper/collections/_children` to list children for a node, and let pure `GET <path>` calls always return UTF-8 data. Futher, it must be possible to retrieve the metadata of any znode. The existing `/admin/zookeeper` API returns both metadata (ctime etc), utf-8 data and children in the same response. So consider either a `GET /api/cluster/zookeeper/collections/_metadata` call, or putting the metadata in HTTP headers `X-solr-zkmeta-ctime: foo`?

An example to show the problem is the `/collections` znode which happens to have no children on a clean node after install. Try this:

    docker run --rm -ti solr:8.4.1 bash
    solr start -c
    curl "http://localhost:8983/solr/admin/zookeeper?path=/collections&detail=true"

It will return `children_count:0`. Your current impl would just return the data of that node, i.e. an empty string, but after creating the first collection it would return a list, and you'd need to add `leaf=true` to get to the data, which is very inconsistent.

File: solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java (hunk 4 above)
Comment (tflobbe) [reply solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java]: This try/catch is inside a loop that makes requests to ZooKeeper. So you are saying that if the thread is interruped (shutdown, or in tests for example) the plan is to swallow that interruption and continue processing (doing requests to ZooKeeper for every child this zkpath has, with sleeps/retries), and the intention of that is to add an "exception" to one (just one) of the elements in the tree in the response?
Regarding other exceptions, lets say you are getting back a SessionExpiredException, the plan is to continue iterating over all the child docs and continue adding `node -> KeeperException.SessionExpiredException`?
I think the proper handling of most these exceptions is to stop processing and return an error. If you get a NoNodeException, then yes, that could be added in the response if you want, but doing a blind `catch (Exception e)` is a bad idea.

File: solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java (hunk 6 above)
Comment (tflobbe) [reply solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java]: It does. See my comment above

Comment (tflobbe) [reply solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java]: Also, wouldn't this be returning 200 OK for error cases?

File: solr/core/src/java/org/apache/solr/handler/admin/ZookeeperRead.java [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+/**Exposes the content of the Zookeeper
+ * This is an expert feature that exposes the data inside the back end zookeeper.This API may change or
+ * be removed in future versions.
+ * This is not a public API. The data that is returned is not guaranteed to remain same
+ * across releases, as the data stored in Zookeeper may change from time to time.
+ */
+@EndPoint(path = "/cluster/zk/*",
+    method = SolrRequest.METHOD.GET,
+    permission = COLL_READ_PERM)
+public class ZookeeperRead {
```
Comment (tflobbe) [inline solr/core/src/java/org/apache/solr/handler/admin/ZookeeperRead.java]: Handlers of requests are typically called `Handler`

File: solr/core/src/test/org/apache/solr/handler/admin/ZookeeperStatusHandlerTest.java (hunk 10 above)
Comment (tflobbe) [reply solr/core/src/test/org/apache/solr/handler/admin/ZookeeperStatusHandlerTest.java]: Wait, can't you do `"this is some test content".getBytes(StandardCharsets.UTF_8);`?

File: solr/core/src/test/org/apache/solr/handler/admin/ZookeeperReadTest.java [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+      o = Utils.executeGET(client.getHttpClient(),
+          basezk + "/configs",
+          Utils.JSONCONSUMER);
+      assertEquals("0", String.valueOf(getObjectByPath(o, true, split(":/configs:_default:dataLength", ':'))));
+      assertEquals("0", String.valueOf(getObjectByPath(o, true, split(":/configs:conf:dataLength", ':'))));
+
+      o = Utils.executeGET(client.getHttpClient(),
+          basezk + "/configs?leaf=true",
+          Utils.JSONCONSUMER);
+      assertTrue(((Map)o).containsKey("/configs"));
```
Comment (tflobbe) [inline solr/core/src/test/org/apache/solr/handler/admin/ZookeeperReadTest.java]: can you fix the compiler warnings?

File: solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java (hunk 2 above)
Comment (tflobbe) [reply solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java]: Ah, true. The problem is that, by requiring the CoreContainer you are making this much more difficult to test (which is why I guess you are starting a Solr cluster just to test this handler). Maybe there is a way to improve the design to improve tests

File: solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java (hunk 5 above)
Comment (tflobbe) [reply solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java]: Is this only the case for the situation where we write the data? what about the case of the child nodes? can those requests be written in other response formats? If this is the only format that's allowed, should we fail if another format is requested?

File: solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java (hunk 4 above)
Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java]: Please take a look at the latest

* NoNodeException returns with a `404`
* Other Exceptions return with `500`

This is supposed to just fail fast if there is a problem. It does not need to be any more sophisticated than that

**Comment by noblepaul** [comment]
> Not sure I like the magic switch between returning actual data and returning a list of metdatata simply based on whether a znode happens to have children or not. This is not deterministic. I'd rather have an explicit action, e.g. `GET /api/cluster/zookeeper/collections/_children` to list children for a node, and let pure `GET <path>` calls always return UTF-8 data. Futher, it must be possible to retrieve the metadata of any znode. The existing `/admin/zookeeper` API returns both metadata (ctime etc), utf-8 data and children in the same response. So consider either a `GET /api/cluster/zookeeper/collections/_metadata` call, or putting the metadata in HTTP headers `X-solr-zkmeta-ctime: foo`?
> 
> An example to show the problem is the `/collections` znode which happens to have no children on a clean node after install. Try this:
> 
> ```
> docker run --rm -ti solr:8.4.1 bash
> solr start -c
> curl "http://localhost:8983/solr/admin/zookeeper?path=/collections&detail=true"
> ```
> 
> It will return `children_count:0`. Your current impl would just return the data of that node, i.e. an empty string, but after creating the first collection it would return a list, and you'd need to add `leaf=true` to get to the data, which is very inconsistent.

I changed it to have 2 paths
* `/api/cluster/zk-data`
* `/api/cluster/zk-ls` 

There is no reason for ambiguity

File: solr/core/src/test/org/apache/solr/handler/admin/ZookeeperReadTest.java (hunk 12 above)
Comment (noblepaul) [reply solr/core/src/test/org/apache/solr/handler/admin/ZookeeperReadTest.java]: what are the warnings?

File: solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java (hunk 6 above)
Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java]: this is fixed

**Comment by noblepaul** [comment]
I plan to commit this soon

**Comment by janhoy** [comment]
Make sure to announce your intention in the JIRA issue too, and allow a few days lazy consensus.

**Comment by tflobbe** [comment]
@noblepaul, seems to me that there are still some unaddressed comments

File: solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java (hunk 4 above)
Comment (tflobbe) [reply solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java]: This is still a problem in your latest code

File: solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java (hunk 2 above)
Comment (tflobbe) [reply solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java]: Have you given any thoughts to this? It's unfortunate to have a feature that has so little (nothing really) to do with other parts of Solr to have to start a full SolrCloud cluster to be able to test.

File: solr/core/src/test/org/apache/solr/handler/admin/ZookeeperStatusHandlerTest.java (hunk 10 above)
Comment (tflobbe) [reply solr/core/src/test/org/apache/solr/handler/admin/ZookeeperStatusHandlerTest.java]: This is still unresolved. Can you clarify why we need this random byte array for the test?

File: solr/core/src/test/org/apache/solr/handler/admin/ZookeeperReadTest.java (hunk 12 above)
Comment (tflobbe) [reply solr/core/src/test/org/apache/solr/handler/admin/ZookeeperReadTest.java]: Isn't your compiler complaining about non-parametrized map?

File: solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java (hunk 4 above)
Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java]: I fail to see the problem. Could you please explain

File: solr/core/src/java/org/apache/solr/handler/admin/ZookeeperRead.java (hunk 11 above)
Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/handler/admin/ZookeeperRead.java]: APIs are named as *API

File: solr/core/src/test/org/apache/solr/handler/admin/ZookeeperReadTest.java (hunk 12 above)
Comment (noblepaul) [reply solr/core/src/test/org/apache/solr/handler/admin/ZookeeperReadTest.java]: NO. I couldn't see it. Maybe you can just commit that directly

File: solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java (hunk 4 above)
Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java]: must be fixed now

Comment (tflobbe) [reply solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java]: Is not yet fixed in the list case

File: solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java (hunk 5 above)
Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java]: no if you are requesting data , you should expect raw data. it will not honour the `wt` param

File: solr/core/src/test/org/apache/solr/handler/admin/ZookeeperStatusHandlerTest.java (hunk 10 above)
Comment (noblepaul) [reply solr/core/src/test/org/apache/solr/handler/admin/ZookeeperStatusHandlerTest.java]: I wanted a big enough `byte[]` not a small one.

File: solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java (hunk 2 above)
Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java]: We should probably  support a `Supplier<ZkClient>`.

File: solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java (hunk 4 above)
Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java]:
```
  public void list(SolrQueryRequest req, SolrQueryResponse rsp) {
      String path = req.getPathTemplateValues().get("*");
      if (path == null || path.isEmpty()) path = "/";
      try {
        List<String> l = coreContainer.getZkController().getZkClient().getChildren(path, null, false);
        String prefix = path.endsWith("/") ? path : path + "/";
        rsp.add(path, (MapWriter) ew -> {
          for (String s : l) {
            Stat stat = null;
            try {
              stat = coreContainer.getZkController().getZkClient().exists(prefix + s, null, false);
            } catch (Exception e) {
              throw new RuntimeException(e);
            }
            printStat(ew, s, stat);
          }
        });
      } catch (KeeperException.NoNodeException e) {
        throw new SolrException(SolrException.ErrorCode.NOT_FOUND, "No such node :"+ path);
      } catch (Exception e) {
        rsp.add(CONTENT, new ContentStreamBase.StringStream(Utils.toJSONString(Collections.singletonMap("error", e.getMessage()))));
      } finally {
        RequestHandlerUtils.addExperimentalFormatWarning(rsp);
      }
    }
```

If there is an exception you will see a 500 error in this code. Am I missing something?

File: solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java (hunk 5 above)
Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java]: yes, for list operations , you can get data in any format
